File: go/148.sort-list.go
/*
 * @lc app=leetcode id=148 lang=golang
 *
 * [148] Sort List
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil { return head }

	dummy := &ListNode{
		Next: head,
	}
	prev := dummy
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		prev = prev.Next
		slow = slow.Next
		fast = fast.Next.Next
	}
	if(slow != nil) {
		prev.Next = nil
	}

	left := sortList(head)
	mid := sortList(slow)

	d := merge(left, mid)

	return d.Next
}

func merge(first, second *ListNode) *ListNode {
	dummy := &ListNode{}
	curr := dummy
	for {
		if second == nil {
			curr.Next = first
			break
		} else if first == nil {
			curr.Next = second
			break
		} else if first.Val < second.Val {
			curr.Next = first
			first = first.Next
		} else {
			curr.Next = second
			second = second.Next
		}

		curr = curr.Next
	}

	return dummy
}

File: java/148.sort-list.java
/* -----------------------------------
 *  WARNING:
 * -----------------------------------
 *  Your code may fail to compile
 *  because it contains public class
 *  declarations.
 *  To fix this, please remove the
 *  "public" keyword from your class
 *  declarations.
 */


/*
 * @lc app=leetcode id=148 lang=java
 *
 * [148] Sort List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode pivot = head;
        ListNode leftHead = new ListNode(-1);
        ListNode rightHead = new ListNode(-1);

        ListNode node = head.next;
        ListNode L = leftHead, R = rightHead, P = pivot;
        while(node != null) {
            if(node.val < pivot.val) {
                L.next = node;
                L = L.next;
            } else if(node.val > pivot.val){
                R.next = node;
                R = R.next;
            } else {
                P.next = node;
                P = P.next;
            }
            node = node.next;
        }
        L.next = null;
        R.next = null;
        
        ListNode sortedLeft = sortList(leftHead.next);
        ListNode sortedRight = sortList(rightHead.next);

        ListNode dummy = new ListNode(-1);
        if(sortedLeft == null) {
            dummy.next = pivot;
        } else {
            dummy.next = sortedLeft;
            ListNode e = sortedLeft;
            while (e != null && e.next != null) {
                e = e.next;
            }
            e.next = pivot;
        }
        
        P.next = sortedRight;

        return dummy.next;
    }
}


// @lc code=end

